src/librustc/ty/print/pretty.rs

```diff
@@ -1532,12 +1536,103 @@ define_print_and_forward_display! {
     }
 
     &'tcx ty::Const<'tcx> {
-        match self.val {
-            ConstValue::Unevaluated(..) |
-            ConstValue::Infer(..) => p!(write("_")),
-            ConstValue::Param(ParamConst { name, .. }) => p!(write("{}", name)),
-            _ => p!(write("{:?}", self)),
+        let u8 = cx.tcx().types.u8;
+        if let ty::FnDef(did, substs) = self.ty.sty {
+            p!(print_value_path(did, substs));
+            return Ok(cx);
+        }
+        if let ConstValue::Unevaluated(did, substs) = self.val {
+            match cx.tcx().describe_def(did) {
+                | Some(Def::Static(_))
+                | Some(Def::Const(_))
+                | Some(Def::AssociatedConst(_)) => p!(print_value_path(did, substs)),
+                _ => if did.is_local() {
+                    let span = cx.tcx().def_span(did);
+                    if let Ok(snip) = cx.tcx().sess.source_map().span_to_snippet(span) {
```

Comment:
I'm not sure I like this, but can you add a test for this?

Reply:
I've been unable to coax MIR to actually hit this

Reply:
Yeah, you can't, you need to rely on the `def-path` testing (see `ui/symbol-names`), I think, and then have `impl Foo for [u8; 1 + 2]` or similar - the `1 + 2` would probably end up in the output.

Or cause a cycle error, those also print paths.

Reply:
def-path of paths with arrays just print `_` for the length

Reply:
(I added a test for that)